File: backend/health-check/src/handler.ts
import { Signer } from '@aws-sdk/rds-signer'
import type {
  APIGatewayProxyEventV2,
  APIGatewayProxyResultV2,
} from 'aws-lambda'
import { Client } from 'pg'

type HealthBody = {
  status: 'ok' | 'error'
  database?: string
  message?: string
}

function json(
  statusCode: number,
  body: HealthBody,
): APIGatewayProxyResultV2 {
  return {
    statusCode,
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(body),
  }
}

function requiredEnv(name: string): string {
  const value = process.env[name]
  if (!value) {
    throw new Error(`Missing required env var: ${name}`)
  }
  return value
}

export async function handler(
  _event: APIGatewayProxyEventV2,
): Promise<APIGatewayProxyResultV2> {
  const host = requiredEnv('DB_HOST')
  const port = Number(process.env.DB_PORT ?? '5432')
  const database = requiredEnv('DB_NAME')
  const user = requiredEnv('DB_IAM_USER')
  const region = requiredEnv('AWS_REGION')

  const signer = new Signer({
    hostname: host,
    port,
    username: user,
    region,
  })

  const client = new Client({
    host,
    port,
    database,
    user,
    password: await signer.getAuthToken(),
    ssl: { rejectUnauthorized: false },
    connectionTimeoutMillis: 5_000,
  })

  try {
    await client.connect()
    await client.query('SELECT 1')
    return json(200, { status: 'ok', database })
  } catch (err) {
    const message = err instanceof Error ? err.message : 'database unavailable'
    console.error('health-check db error:', message)
    return json(503, { status: 'error', message })
  } finally {
    await client.end().catch(() => undefined)
  }
}

File: backend/services/health-check/src/handler.py
import json
import os

from sqlalchemy import text

from shared.db.engine import run_async
from shared.db.session import get_session


def _json(status_code: int, body: dict) -> dict:
    return {
        "statusCode": status_code,
        "headers": {"content-type": "application/json"},
        "body": json.dumps(body),
    }


async def _check_db() -> str:
    database = os.environ["DB_NAME"]
    async with get_session() as session:
        await session.execute(text("SELECT 1"))
    return database


def lambda_handler(_event, _context):
    try:
        database = run_async(_check_db())
        return _json(200, {"status": "ok", "database": database})
    except Exception as exc:
        message = str(exc)
        print(f"health-check db error: {message}")
        return _json(503, {"status": "error", "message": message})
